### DISCUSSION TITLE
Outputting a `file` attribute for the Junit reporter

### DISCUSSION THREAD

**Discussion by roykolak** [vitest-dev/vitest#4792] [Q&A]
I'm trying to feed junit timing data into a spec splitting cli tool (https://github.com/mtsmfm/split-test)

The built-in junit reporter is creating the timing data, however it is not generating a `<testsuite>` with a `file` attribute. Is there an easy way to get this attribute to be included w/o creating a custom reporter?

**Comment by hi-ogawa** [comment]
I think currently Vitest's junit reporter includes file name in `<testsuite name="...">` and `<testcase classname="...">`. 
Just to make sure we are talking about the same thing, this is what I see when I use junit reporter for the tests in https://github.com/vitest-dev/vitest/tree/main/test/core/test

<details><summary>Reveal junit reporter output</summary>

```xml
<?xml version="1.0" encoding="UTF-8" ?>
<testsuites name="vitest tests" tests="1349" failures="0" errors="0" time="6.725">
    <testsuite name="test/basic.test.ts" timestamp="2023-12-23T04:14:34.684Z" hostname="myhostname" tests="9" failures="0" errors="0" skipped="1" time="0.111">
        <testcase classname="test/basic.test.ts" name="Math.sqrt()" time="0.002">
        </testcase>
        <testcase classname="test/basic.test.ts" name="JSON" time="0.001">
        </testcase>
        <testcase classname="test/basic.test.ts" name="mode and NODE_ENV is test by default" time="0.001">
        </testcase>
         ...
    </testsuite>
    <testsuite name="test/circular.test.ts" timestamp="2023-12-23T04:14:34.691Z" hostname="myhostname" tests="2" failures="0" errors="0" skipped="0" time="0.106">
        <testcase classname="test/circular.test.ts" name="circular" time="0.001">
        </testcase>
        <testcase classname="test/circular.test.ts" name="timeout" time="0.1">
        </testcase>
    </testsuite>
    ...
```

</details>

I think it's worth noting that either `testcase.file` nor `testcase.filepath` doesn't seem to be junit standard as mentioned in https://github.com/mtsmfm/split-test?tab=readme-ov-file#note

> To be exact it isn't JUnit standard.

For the time being, maybe just replacing `name=` with `filepath=` could work? Most primitive way would be something like:

```sh
sed 's/<testsuite name=/<testsuite filepath=/' junit.xml 
...
    <testsuite filepath="test/basic.test.ts" timestamp="2023-12-23T04:14:34.684Z" hostname="myhostname" tests="9" failures="0" errors="0" skipped="1" time="0.111">
```

Not sure Vitest team would consider changing this, but if this is about making it more compatible with Jest ecosystem, then it's probably worth raising an issue at least (for example, there was feature request like this before https://github.com/vitest-dev/vitest/issues/3691).

**Comment by roykolak** [comment]
Thanks for the response @hi-ogawa ! I ended up doing something similar (using xmlstarlet) to just edit the generated junit report with a new `file` attribute containing the full path to the spec file... 

```
xmlstarlet ed -a '//testsuite[@name]' -t attr -n file test-results.xml | xmlstarlet ed -u '//testsuite/@file' -x "concat(\"$(pwd)/\", ../@name)" > test-results-modified.xml
```

Will plan on raising an issue. Totally get that this is a non standard attribute, so it makes sense why the vitest junit reporter is not outputting it. However [mocha-junit-reporter](https://github.com/michaelleeallen/mocha-junit-reporter#readme) is outputting this attribute by default and circle ci is also looking for a `file` attribute in junit files so it seems to be a community standard
